stan/math/mix.hpp

```diff
@@ -2,11 +2,19 @@
 #define STAN_MATH_MIX_HPP
 
 #include <stan/math/mix/meta.hpp>
-
-#include <stan/math/prim.hpp>
-#include <stan/math/fwd.hpp>
-#include <stan/math/rev.hpp>
 #include <stan/math/mix/fun.hpp>
 #include <stan/math/mix/functor.hpp>
 
+#include <stan/math/rev/core.hpp>
+#include <stan/math/rev/meta.hpp>
+#include <stan/math/rev/fun.hpp>
+#include <stan/math/rev/functor.hpp>
+
+#include <stan/math/fwd/core.hpp>
+#include <stan/math/fwd/meta.hpp>
+#include <stan/math/fwd/fun.hpp>
+#include <stan/math/fwd/functor.hpp>
+
+#include <stan/math/prim.hpp>
```

Comment:
Ahh, there must be more to the story. Do tell!

Reply:
Same for rev.

Reply:
And why the individual files?

Reply:
@syclik do you have any memories of why prim was originally included before `rev` and `fwd`? `prim` is the most general, so it makes sense to have it last here (see also [this](https://github.com/stan-dev/math/pull/2190/files#r536395799) comment from this pull), but there could be something else I don't want to accidentally miss.

Reply:
rev.hpp for example also includes prim, which we dont want included before fwd.

Reply:
Good point.

Reply:
Yes, the reason we had `prim` included before `rev` and `fwd` was because `prim` provided all the primary template definitions. Each function in `rev` and `fwd` could template specialize those functions.

To avoid duplicate definitions, we used the `meta` include pretty cleverly, which may have been too clever.

@t4c1, I'm guessing the original include order caused a cascade of compiler errors? If that's the case, one of the changes may have inadvertently brought in too much. (I've done this countless times and had to work through it.) If you did want to detangle it, you can use the generated `.d` files or use the compiler to generate the dependency graph for the compilation process to try to track it down.

Reply:
The way I changed it in this PR seem to work, so I feel no need for extra modifications.

Reply:
I cloned this math and verified that it works in a Cmdstan, so since it's not obviously broken and it seems to work at least that far upstream, I'm okay with it.

Reply:
@syclik thanks for taking a look

Reply:
Actually the logic for the change makes sense but why now? Nothing really changed in this pull request in terms of new functions and what not. Gonna flip these includes back real quick.

We are presumably solving a problem with ADL here. Like, the easiest way to make sure a function is found is to define it before it is used, but that isn't the only way a function is found. For instance, if you call:

```
stan::math::var a;
myfunc(a);
```

I think the entire namespace `stan::math` is searched for a definition of `myfunc`, not just the stuff defined before this bit of code.

Reply:
Well, the last thing I posted is wrong regarding ADL.

But since I'm playing with this I'm not able to get this to cause any errors.

But I'm also more convinced by what I see the only way this is wrong is that mix is included before rev and fwd and mix might reasonably depend on those lol.

Reply:
@t4c1 presumably this blew up in one of the Jenkins tests here. You know which?

Reply:
Not to be too pedantic -- I'm not going to spend too much time trying to make this not work but it bears a little investigating.

Reply:
I think it was the combination of these two:
https://jenkins.mc-stan.org/blue/organizations/jenkins/Math%20Pipeline/detail/PR-2190/15/pipeline
https://jenkins.mc-stan.org/blue/organizations/jenkins/Math%20Pipeline/detail/PR-2190/14/pipeline

Reply:
Oh and I see the conversation back here: https://github.com/stan-dev/math/pull/2190#issuecomment-730683463

Reply:
> But I'm also more convinced by what I see the only way this is wrong is that mix is included before rev and fwd and mix might reasonably depend on those lol.

Thqat can not really be an issue. If mix needs fwd or rev it is simply included. Same if fwd or rev needs prim.

Reply:
I don't remember what exactly was an issue, but it was something, that gcc allows even if it is not standard, while clang is more strict. I know I was reading this: https://clang.llvm.org/compatibility.html#dep_lookup

Reply:
Yeah I think this change is right.

If everything is template overloads, then we can define things after they are used. Like in this: https://godbolt.org/z/hbT54n

(you'll see both `test::myfunc` functions get called)

If you add an overload of myfunc, it will only get called if the overload is already available when the compiler hits `whatever`. For instance, you can add this code above and below `whatever` and see what happens:

```
int myfunc(const char *) {
    std::cout << "hey" << std::endl;
    return 5;
}
```

Since we're depending on overloading and templates in Stan math, we gotta make sure and include the overloads for the `rev` and `fwd` types before the more general `prim` functions or the stuff in `prim` might not be able to find everything.